cleverhans/attacks_eager.py

```diff
@@ -0,0 +1,331 @@
+
+    def generate_np(self, x_val, **kwargs):
+        """
+        Generate adversarial examples and return them as a NumPy array.
+
+        :param x_val: A NumPy array with the original inputs.
+        :param **kwargs: optional parameters used by child classes.
+        :return: A NumPy array holding the adversarial examples.
+        """
+        tfe = tf.contrib.eager
+        x = tfe.Variable(x_val)
+        adv_x = self.generate(x, **kwargs)
+        return adv_x.numpy()
+
+    def clip_eta(self, eta, ord, eps):
+        """
+        Helper function to clip the perturbation to epsilon norm ball.
+        :param eta: A tf variable with the current perturbation.
+        :param ord: Order of the norm (mimics Numpy).
+                    Possible values: np.inf, 1 or 2.
+        :param eps: Epilson, bound of the perturbation.
+        """
+        # Clipping perturbation eta to self.ord norm ball
+        if ord not in [np.inf, 1, 2]:
+            raise ValueError('ord must be np.inf, 1, or 2.')
+        reduc_ind = list(xrange(1, len(eta.get_shape())))
+        avoid_zero_div = 1e-12
+        if ord == np.inf:
+            eta = tf.clip_by_value(eta, -eps, eps)
+        else:
+            if ord == 1:
+                norm = tf.maximum(avoid_zero_div,
+                                  tf.reduce_sum(tf.abs(eta),
+                                                reduc_ind, keep_dims=True))
+            elif ord == 2:
+                # avoid_zero_div must go inside sqrt to avoid a divide by zero
+                # in the gradient through this operation
+                norm = tf.sqrt(tf.maximum(avoid_zero_div,
+                                          tf.reduce_sum(tf.square(eta),
+                                                        reduc_ind,
+                                                        keep_dims=True)))
+            # We must *clip* to within the norm ball, not *normalize* onto the
+            # surface of the ball
+            factor = tf.minimum(1., eps / norm)
+            eta = eta * factor
+        return eta
+
+    def model_loss(self, logits, labels, mean=True):
+        """
+        Cross Entropy Loss
+        :param labels: correct labels
+        :param logits: logits predicted by the model
+        :param mean: boolean indicating whether should return mean of loss
+                     or vector of losses for each input of the batch
+        :return: return mean of loss if True, otherwise return vector with per
+                 sample loss
+        """
+        try:
+            labels = tf.stop_gradient(labels)
+            out = tf.nn.softmax_cross_entropy_with_logits_v2(
+                                    logits=logits, labels=labels)
+        except AttributeError:
+            out = tf.nn.softmax_cross_entropy_with_logits(
+                                    logits=logits, labels=labels)
+        if mean:
+            out = tf.reduce_mean(out)
+        return out
+
+
+class ListOfAttacks(object):
```

Comment:
1. What is the use case for this?
2. Why use a class instead of say?
ATTACKS = {'fsgm': FastGradientMethod}

Reply:
In cleverhans_tutorials for tf backend each attack has one tutorial, 
but for eager backend we can have one tutorial that accepts --attack as a argument.
mnist_tutorial_eager.py expects  a string for --attack argument, (eg: 'fgsm') and maps it to Attack class 
 (eg: FastGradientMethod).

This class is used for above mapping attack name (string) to Attack class, to reduce code repetition, what are your thoughts on this ?

Reply:
Right but a dictionary is just as easy to import isn't it?
If the list is for the tutorial, may be move it to the tutorial.
Another problem with the list, is that it has to be manually maintained, someone adding a new attack could forget to put it in the list, although if it's just in the tutorial folder it's okay.

Reply:
True, i will move this class to tutorial, 
dictionary is just as easy to import and serves the purpose, modified in the updated PR.